### PR TITLE
Fix Swift Package Manager compile issue

### PR DESCRIPTION
**Pull request by rynecheow**
Signed-off-by: Ryne Cheow <rynecheow@gmail.com>

### Issue Link :link:
https://github.com/danielgindi/Charts/issues/4016

### Goals :soccer:
1. Resolve dependency errors from SPM builds or `swift build`

### Implementation Details :construction:
Resolve dependencies for correct platform

### Testing Details :mag:
Compile for all platforms using Swift 5.0 toolchain

### CODE REVIEW

**Comment by rynecheow** [comment]
@liuxuan30 Could you help to review this?

**Comment by liuxuan30** [comment]
Hmm this seems caused by Xcode 11 beta? @rynecheow any documentation or reason, about this behavior, that it's expected for Xcode 11, or new Swift version?

Because given the history dealing with swift or xcode upgrade, I have seen the same changes about this import back and forth. What I'm concerning is this might be a bug for Xcode 11 beta, not us. 

Any idea @jjatie ?

I could also prefer holding this PR until Xcode 11 GM is released. 

But, for anyone testing Xcode 11 or iOS13, we could create a branch for it, and you can commit your PR to this branch. Is this ok for you?

**Comment by rynecheow** [comment]
Hey @liuxuan30 I just ran a deeper test using **current** stable tools (Swift 5.0 toolchain and Xcode 10.2.1) and build has failed as well. Updated the issue to reflect more accurate problem. 

Basically `swift build` would fail.

**Comment by liuxuan30** [comment]
@rynecheow it's strange, our CI and my Xcode 10.2.1 just works well.

**Comment by rynecheow** [comment]
@liuxuan30 Currently CI is not building the SPM package according build script / build log I believe.

**Comment by jjatie** [comment]
This shouldn't be merged until later this summer. We are only on beta 1 and things are subject to change (especially where all the SPM stuff was developed in private). I noticed the toolchain wasn't bumped in this PR

**Comment by rynecheow** [comment]
@jjatie This pull request aims to fix the build error on current stable Swift toolchain - not Xcode 11 
re. #4016

**Comment by liuxuan30** [comment]
@rynecheow I kind of understand what you mean now.

 if this is about fixing 'Swift Package Manager' compiling - which is rarely used by me or other people so far (you are the first one reporting this), I would like to know or maybe send a bug to APPL first to figure out why only this `swift compile` fails to build, while `xcodebuild` works well.

It's kind of inconsistent, as SPM fails to find the framework while xcodebuild can handle it?

**Comment by jjatie** [comment]
@rynecheow The current toolchain doesn't support iOS. I'm not sure how we could fix it without migrating to the new toolchain.

**Comment by jjatie** [comment]
Why was Package.swift not updated to the 5.0 toolchain?
Why were the demos touched at all when they aren't part of Package.swift?

**Comment by mattjgalloway** [comment]
Chiming in here - I think that we should commit this. Reason being, things are being used when they're not explicitly imported. I haven't looked into why this works when compiling using the Xcode project, but just take a look at this:

`DataApproximator+N.swift`:

```
import Foundation

extension CGPoint {
    fileprivate func distanceToLine(from linePoint1: CGPoint, to linePoint2: CGPoint) -> CGFloat {
        let dx = linePoint2.x - linePoint1.x
        let dy = linePoint2.y - linePoint1.y
        
        let dividend = abs(dy * self.x - dx * self.y - linePoint1.x * linePoint2.y + linePoint2.x * linePoint1.y)
        let divisor = sqrt(dx * dx + dy * dy)
        
        return dividend / divisor
    }
}
```

How can we be extending `CGPoint` without importing CoreGraphics where `CGPoint` resides? AFAIK Foundation in no way shouldn't be importing CoreGraphics. Must be something about the Xcode project that means that CoreGraphics (or UIKit so CoreGraphics recursively) is imported when compiling this file. But really, we should be importing what we use anyway.

**Comment by RamblinWreck77** [comment]
@liuxuan30 Also confirming this issue still occurs in SwiftPM on the Gold Master of Xcode11.

**Comment by RamblinWreck77** [comment]
Also, the Package.swift specifies:

```
 .library(name: "Charts", type: .dynamic, targets: ["Charts"])
```

While the docs specify it is best to leave it nil and let SwiftPM decide. Is there any reason to force .dynamic? Dynamic libraries impact app boot times on iOS.

**Comment by liuxuan30** [comment]
@jjoelson can you check what @RamblinWreck77 said?

I'm open to merge this, if the final compiler takes it this way.

File: Source/Supporting Files/Info.plist [hunk 1]
```diff
@@ -15,7 +15,7 @@
 	<key>CFBundlePackageType</key>
 	<string>FMWK</string>
 	<key>CFBundleShortVersionString</key>
-	<string>3.3.0</string>
+	<string>3.3.1</string>
```
Comment (liuxuan30) [inline Source/Supporting Files/Info.plist]: please don't touch it in this PR

File: ChartsDemo-iOS/Swift/Demos/CombinedChartViewController.swift [hunk 2]
```diff
@@ -40,11 +41,11 @@ class CombinedChartViewController: DemoBaseViewController {
         chartView.highlightFullBarEnabled = false
         
         
-        chartView.drawOrder = [DrawOrder.bar.rawValue,
-                               DrawOrder.bubble.rawValue,
-                               DrawOrder.candle.rawValue,
-                               DrawOrder.line.rawValue,
-                               DrawOrder.scatter.rawValue]
+        chartView.drawOrder = [CombinedChartView.DrawOrder.bar.rawValue,
```
Comment (liuxuan30) [inline ChartsDemo-iOS/Swift/Demos/CombinedChartViewController.swift]: why change this?

**Comment by liuxuan30** [comment]
https://github.com/danielgindi/Charts/pull/4118
seems this is ready to merge, if you don't want to commit more?

File: ChartsDemo-iOS/Swift/Demos/CombinedChartViewController.swift (hunk 2 above)
Comment (rynecheow) [reply ChartsDemo-iOS/Swift/Demos/CombinedChartViewController.swift]: Due to the ambiguity of the enum

File: Source/Supporting Files/Info.plist (hunk 1 above)
Comment (rynecheow) [reply Source/Supporting Files/Info.plist]: I'll revert.

File: ChartsDemo-iOS/Swift/Demos/CombinedChartViewController.swift (hunk 2 above)
Comment (rynecheow) [reply ChartsDemo-iOS/Swift/Demos/CombinedChartViewController.swift]: Changed back to original - my bad here!

**Comment by jjoelson** [comment]
@liuxuan30 The SwiftPM docs do say to leave `type` nil unless the library disallows a linkage type for other reasons:

>Leave this parameter unspecified to let to let the Swift Package Manager choose between static or dynamic linking (recommended). If you do not support both linkage types, use `.static` or `.dynamic` for this parameter.

But then there's also an example that has separate products for static and dynamic, so it's a bit confusing:
```
let package = Package(
    name: "Paper",
    products: [
        .executable(name: "tool", targets: ["tool"]),
        .library(name: "Paper", targets: ["Paper"]),
        .library(name: "PaperStatic", type: .static, targets: ["Paper"]),
        .library(name: "PaperDynamic", type: .dynamic, targets: ["Paper"]),
    ],
    dependencies: [
        .package(url: "http://example.com.com/ExamplePackage/ExamplePackage", from: "1.2.3"),
        .package(url: "http://some/other/lib", .exact("1.2.3")),
    ],
    targets: [
        .target(
            name: "tool",
            dependencies: [
                "Paper",
                "ExamplePackage"
            ]),
        .target(
            name: "Paper",
            dependencies: [
                "TSCBasic",
                .target(name: "Utility"),
                .product(name: "AnotherExamplePackage"),
            ])
    ]
)
```

**Comment by liuxuan30** [comment]
there is no reason to only use dynamic, but by default, this "library" is built with framework, so I guess that's why it's dynamic. I think we should be fine with nil type, as long as SPM has no bugs handling it.

@rynecheow can you also check what #4118 did for `Package.swift `? Basically I would like to merge this as you are the first PR. I didn't see you touch this file, but #4118 change a few, especially 
>.target(name: "Charts", dependencies: [], path: "./Source", publicHeadersPath: "./Source/Supporting Files")

I'm not sure the impact about this, as I'm not following SPM trend now. Besides, I will be grateful if you could modify the type to nil as discussed above, so I don't have to pull your branch. Thanks.

File: .gitignore [hunk 3]
```diff
@@ -72,3 +72,4 @@ fastlane/test_output
 Carthage
 Charts.framework.zip
 ChartsRealm.framework.zip
+.swiftpm
```
Comment (liuxuan30) [inline .gitignore]: `.swiftpm` or `.swiftpm/`?

Comment (rynecheow) [reply .gitignore]: Rightfully `.swiftpm/` - let me change that. Did `.swiftpm` cause it does the trick.

**Comment by rynecheow** [comment]
> there is no reason to only use dynamic, but by default, this "library" is built as framework, so I guess that's why it's dynamic. I think we should be fine with nil type, as long as SPM has no bugs handling it.
> 
> @rynecheow can you also check what #4118 did for `Package.swift `? Basically I would like to merge this as you are the first PR. I didn't see you touch this file, but #4118 change a few, especially
> 
> > .target(name: "Charts", dependencies: [], path: "./Source", publicHeadersPath: "./Source/Supporting Files")
> 
> I'm not sure the impact about this, as I'm not following SPM trend now. Besides, I will be grateful if you could modify the type to nil as discussed above, so I don't have to pull your branch. Thanks.

`path: "./Source"` is unnecessary as it is the parameter automatically defaults to a certain pattern if it is not specified. 
https://developer.apple.com/documentation/swift_packages/target/2880335-path

`publicHeadersPath: "./Source/Supporting Files"` I am not sure if this is completely necessary as I believe SPM already handles that at its build process.

**Comment by liuxuan30** [comment]
Seems green light now. BTW, do you think you can add script code for travis to build via SPM? Just to build it to make sure it builds fine, No UT is needed.

Otherwise I will merge next Monday.

**Comment by liuxuan30** [comment]
A second thought we should just go ahead, if you think travis ci is needed, let's add later. Not surehowpopular spm can be

**Pull request merged**